normandy/recipes/exports.py

```diff
@@ -0,0 +1,172 @@
+                raise ImproperlyConfigured(msg)
+
+        # Test authentication.
+        server_info = self.client.server_info()
+        is_authenticated = (
+            "user" in server_info
+            and settings.REMOTE_SETTINGS_USERNAME in server_info["user"]["id"]
+        )
+        if not is_authenticated:
+            raise ImproperlyConfigured("Invalid Remote Settings credentials")
+
+        # Test that collection is writable.
+        metadata = self.client.get_collection()
+        if server_info["user"]["id"] not in metadata["permissions"].get("write", []):
+            raise ImproperlyConfigured(
+                f"Remote Settings collection {self.collection_id} is not writable."
+            )
+
+        # Test that collection has the proper review settings.
+        capabilities = server_info["capabilities"]
+        if "signer" in capabilities:
+            signer_config = capabilities["signer"]
+            normandy_resource = [
+                r
+                for r in signer_config["resources"]
+                if r["source"]["bucket"] == settings.REMOTE_SETTINGS_BUCKET_ID
+                and r["source"]["collection"] == self.collection_id
+            ]
+            review_disabled = (
+                len(normandy_resource) == 1
+                and not normandy_resource[0].get(
+                    "to_review_enabled", signer_config["to_review_enabled"]
+                )
+                and not normandy_resource[0].get(
+                    "group_check_enabled", signer_config["group_check_enabled"]
+                )
+            )
+            if not review_disabled:
+                raise ImproperlyConfigured(
+                    f"Review was not disabled on Remote Settings collection {self.collection_id}."
+                )
+
+    def publish(self, recipe):
+        """
+        Publish the specified `recipe` on the remote server by upserting a record.
+        """
+        if self.client is None:
+            return  # no-op if disabled.
+
+        # 1. Put the record.
+        record = recipe_as_record(recipe)
+        self.client.update_record(data=record)
+        try:
+            # 2. Approve the changes immediately (multi-signoff is disabled).
+            self.client.patch_collection(id=self.collection_id, data={"status": "to-sign"})
+
+        except kinto_http.exceptions.KintoException:
+            # Approval failed. Cancel changes and raise.
+            try:
+                # Restore the record already published.
+                in_prod = self.client.get_record(id=record["id"], bucket=MAIN_BUCKET_ID)
+                self.client.update_record(data=in_prod)
+            except kinto_http.exceptions.KintoException as e:
+                # Record was never published, delete it.
+                if e.response.status_code == 404:
+                    self.client.delete_record(id=record["id"])
+            raise
```

Comment:
This is the wrong exception. This will potentially raise the second exception.

Reply:
How could this be the second exception? A bare `raise` in an exception handler always raises the exception that triggered the `except` block. I tested it here:

```python
>>> def foo():
...   try:
...     raise Exception("first")
...   except:
...     try:
...       raise Exception("second")
...     except:
...       pass
...     raise
... 
>>> foo()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in foo
Exception: first
```

Reply:
Great. I had old Python 2.7 in my head. I remember Socorro having this and it was really wrong. Socorro was/is 2.7.